### PR TITLE
Update wasmtime and wit-bindgen 

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR updates the `wasmtime` dependency and the
`wit-bindgen-guest-rust` dependencies to namely update the `wit-parser`
and component underlying implementations. This pulls in
https://github.com/bytecodealliance/wasm-tools/pull/867 which is the implementation of `use` for
WIT.

This PR also refactors the monolithic `wasi.wit` file into multiple
files. I've taken some liberties in naming here so suggestions are
definitely welcome! I've resolved some TODO annotations as well about
`use`-ing types between interfaces. There are two issues remaining,
however:

* The `wasi-command` world still hardcodes `u32` because `world` items
  don't support `use` just yet. That isn't a hard limitation of WIT,
  however, it's just a temporary limitation of the implementation.

* The `wasi-clocks` interface defines the `wasi-future` type. This is
  due to a cyclic dependency where `wasi-future` is defined within
  `wasi-poll` which depends on `wasi-clocks` for types. I've left this
  to a future refactoring to probably have a `types` interface of some
  form somewhere.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Note that this is a draft until https://github.com/bytecodealliance/wasmtime/pull/5622 and https://github.com/bytecodealliance/wasm-tools/pull/885 have merged as well.

**Comment by alexcrichton** [comment]
Those two PRs were merged but I [found more bugs](https://github.com/bytecodealliance/wasm-tools/pull/886) and CI won't work until I publish new versions of `wasm-tools` crates on crates.io. I've rebased on `main`, however.

**Comment by alexcrichton** [comment]
Ok should be good to go now!

File: Cargo.toml [hunk 1]
```diff
@@ -60,3 +60,7 @@ overflow-checks = false
 
 [features]
 command = []
+
+[patch.crates-io]
+wit-parser = { git = 'https://github.com/alexcrichton/wasm-tools', branch = 'more-fuzz-bugs' }
+wit-component = { git = 'https://github.com/alexcrichton/wasm-tools', branch = 'more-fuzz-bugs' }
```
Comment (sunfishcode) [inline Cargo.toml:66]: Can we drop these crate overrides now?

Comment (alexcrichton) [reply Cargo.toml]: Oops, indeed!

**Pull request merged**